[py] Remove redundant driver_instance from conftest.py

lauromoura opened the pull request:
### **User description**
### 🔗 Problem description
Currently, the driver lifetime is handled by two global variables:

- `driver_instance`, the actual Selenium API driver.
- `selenium_driver`, the wrapper `conftest.Driver`, which "owns" the former.

Also, conftest.py could quit the driver from multiple places:

- `Driver.stop_driver` finalizer, used for bidi and xfail tests
- `stop_driver` Session-scope fixture finalizer
- `pytest_exception_interact` to handle exceptions

The last two paths called `driver_instance.quit()` directly but did not tell `selenium_driver` that the driver had exited. This could potentially raise `urllib.exception.MaxRetryError`, as `Driver.stop_driver` would try to send the `quit()` command to an already destroyed driver service.

For example, this happened rather frequently on WebKit import of selenium 4.34.2 tests, whenever an exception happened, as we have enabled the `bidi` flag when calling pytest.

### 💥 What does this PR do?

To address this issue, this commit removes the global variable `driver_instance`, keeping only `selenium_driver` as global, and routes all teardown paths through `Driver.stop_driver()`.

### 🔄 Types of changes
<!-- ✂️ Please delete anything that doesn't apply -->
- Cleanup (formatting, renaming)
- Bug fix (backwards compatible)


___

### **PR Type**
Bug fix, Other


___

### **Description**
- Remove redundant `driver_instance` global variable from conftest.py

- Route all driver teardown paths through `Driver.stop_driver()` method

- Fix potential `MaxRetryError` when driver quit called multiple times

- Improve driver lifecycle management consistency


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Multiple teardown paths"] --> B["driver_instance.quit()"]
  A --> C["selenium_driver.stop_driver()"]
  B --> D["Potential MaxRetryError"]
  E["Single teardown path"] --> F["selenium_driver.stop_driver()"]
  F --> G["Clean driver shutdown"]
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>conftest.py</strong><dd><code>Consolidate driver teardown logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

py/conftest.py

<ul><li>Remove global <code>driver_instance</code> variable declaration<br> <li> Convert <code>stop_driver</code> from property to method in Driver class<br> <li> Update all teardown paths to use <code>selenium_driver.stop_driver()</code><br> <li> Add null check for <code>selenium_driver</code> in BiDi test cleanup</ul>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16271/files#diff-98e9e290ede5d0c7ac9e7df5b49edf63f5df5fcc946b7736e4a6139e4cda26e3">+18/-26</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

lauromoura committed:
[py] Remove redundant driver_instance from conftest.py

Currently, the driver lifetime is handled by two global variables:

- `driver_instance`, the actual Selenium API driver.
- `selenium_driver`, the wrapper `conftest.Driver`, which "owns" the
  former.

Also, conftest.py could quit the driver from multiple places:

- `Driver.stop_driver` finalizer, used for bidi and xfail tests
- `stop_driver` Session-scope fixture finalizer
- `pytest_exception_interact` to handle exceptions

The last two paths called `driver_instance.quit()` directly but did
not tell `selenium_driver` that the driver had exited. This could
potentially raise `urllib.exception.MaxRetryError`, as
`Driver.stop_driver` would try to send the `quit()` command to an
already destroyed driver service.

For example, this happened rather frequently on WebKit import of
selenium 4.34.2 tests, whenever an exception happened, as we have
enabled the `bidi` flag when calling pytest.

To address this issue, this commit removes the global variable
`driver_instance`, keeping only `selenium_driver` as global,
and routes all teardown paths through `Driver.stop_driver()`.

cgoldberg commented:
Thanks @lauromoura 
Can you add the "Properly stop driver before nullifying" suggestion from the review bot? https://github.com/SeleniumHQ/selenium/pull/16271#issuecomment-3235777156

lauromoura committed:
Properly stop driver before nullifying

As suggested by the review bot

cgoldberg commented:
This is causing a failure in CI.

https://github.com/SeleniumHQ/selenium/actions/runs/17329416928/job/49202234767?pr=16271

lauromoura commented:
> This is causing a failure in CI.
> 
> https://github.com/SeleniumHQ/selenium/actions/runs/17329416928/job/49202234767?pr=16271

Checking it. Thanks for the heads-up.

lauromoura committed:
[py] Ignore exception in no_driver_after_test fixture

The `no_driver_after_test` indicates that tests are expected
to quit the driver themselves, leaving `selenium_driver` with a stale
driver instance.

This commit keeps the `stop_driver()` call as a safety fallback, but
ignores WebDriverExceptions since they're expected in the call when
the test properly quits the driver.

lauromoura commented:
@cgoldberg CI fixed :)

cgoldberg reviewed (approved):
LGTM. Thanks @lauromoura
